Sources/PackageLoading/PackageBuilder.swift

```diff
@@ -1383,13 +1383,28 @@ private extension Manifest {
 
 extension Sources {
     var hasSwiftSources: Bool {
-        paths.first?.extension == "swift"
+        return paths.contains { path in
+            guard let ext = path.extension else { return false }
+
+            return FileRuleDescription.swift.fileTypes.contains(ext)
+        }
     }
 
     var containsMixedLanguage: Bool {
-        let swiftSources = relativePaths.filter{ $0.extension == "swift" }
-        if swiftSources.isEmpty { return false }
-        return swiftSources.count != relativePaths.count
+        let supportedClangFileExtensions = FileRuleDescription.clang.fileTypes.union(FileRuleDescription.asm.fileTypes)
+        let supportedSwiftFileExtensions = FileRuleDescription.swift.fileTypes
+
+        let clangSources = relativePaths.filter { path in
+            guard let ext = path.extension else { return false }
+            return supportedClangFileExtensions.contains(ext)
+        }
+
+        let swiftSources = relativePaths.filter { path in
+            guard let ext = path.extension else { return false }
+            return supportedSwiftFileExtensions.contains(ext)
+        }
+
```

Comment:
since we may have many sources to iterate over, the loops above could return earlier- ie as soon as a one file is found

Reply:
Great feedback, thank you. I noticed I can clean up more too and reuse the `hasSwiftSources` property for this. Update incoming...